File: server/src/schema.ts
/**
 * The ID of an entity that exists
 */
export type EntityID = number;

/**
 * The ID of a team.
 * TeamID 0 is always the ID of the non-player team (i.e. the team of hedges).
 */
export type TeamID = number;

/**
 * The ID of a gameID. A UUID v4.
 */
export type GameID = string;

/**
 * The key to a key-protected gameID.
 */
export type PlayerKey = string;

/**
 * A location in the world.
 */
export interface Location {
    x: number;
    y: number;
}

/**
 * The type of an entity.
 */
export type EntityType = "thrower" | "statue" | "hedge";

/**
 * A map tile.
 */
export type MapTile = "G" | "D";

/**
 * The data for an entity in the world.
 */
export interface EntityData {
    id: EntityID;
    type: EntityType;
    location: Location;
    hp: number;
    teamID: TeamID;
    cooldownEnd?: number;
    heldBy?: EntityID;
    holding?: EntityID;
    holdingEnd?: number;
}

/**
 * The data for a Sector
 */
export interface SectorData {
    topLeft: Location;
    controllingTeamID: TeamID;
}

/**
 * The data for the map.
 * 
 * Note: map.json files consist of a MapData object serialized to JSON.
 */
export interface MapFile {
    /**
     * Required sanity check.
     */
    version: "battlecode 2017 hackathon map",

    /**
     * Number of teams required to play the map (not incl. neutral)
     */
    teamCount: number;

    /**
     * The height of the map in tiles.
     */
    height: number;

    /**
     * The width of the map in tiles.
     */
    width: number;

    /**
     * Map is partitioned into square Sectors
     * Each Sector stores Location of its top left corner and covers sectorSize * sectorSize area 
     */
    sectorSize: number;

    /**
     * Slightly wacky indexing:
     * Indexed as [height - (y + 1)][x];
     * This allows a list of tiles to be written out in a text file
     * such that their lower left corner will be placed at (0,0),
     * up is y, and right is x.
     */
    tiles: MapTile[][];

    /**
     * The initial entities on the map.
     */
    entities: EntityData[];

    /**
     * The name of the map.
     * Note: you don't actually need to write this in a map file, it'll just get overwritten when loaded.
     */
    mapName?: string;

}

export interface GameState extends MapFile {
    /** 
     * Sectors are not stored in a map file but are stored in the game state
     * sent to the client.
     */
    sectors: SectorData[];
}

/**
 * Team data.
 */
export interface TeamData {
    teamID: TeamID;

    /**
     * The name of the team.
     */
    name: string;

    /**
     * The key the player used to login to the gameID.
     * unset if the player used no key.
     */
    key?: PlayerKey;
}

/**
 * The neutral team.
 */
export const NEUTRAL_TEAM: TeamData = {
    teamID: 0,
    name: 'neutral',
}

// All possible actions.

export interface MoveAction {
    action: "move";
    id: EntityID;
    dx: number;
    dy: number;
}
export interface PickupAction {
    action: "pickup";
    id: EntityID;
    pickupID: EntityID;
}
export interface ThrowAction {
    action: "throw";
    id: EntityID;
    dx: number;
    dy: number;
}
export interface BuildAction {
    action: "build";
    id: EntityID;
    dx: number;
    dy: number;
}
export interface DisintegrateAction {
    action: "disintegrate";
    id: EntityID;
}

///////////////////////
// Client -> server. //
///////////////////////

/**
 * Request to take part in a gameID.
 */
export interface Login {
    command: "login";

    /**
     * The name of the bot's team.
     * Note, we override this field if the bot is playing on the game server.
     */
    name: string;
    
    /**
     * The gameID the player desires to connect to.
     * If this field is null, the server will make a new gameID and connect the client to it.
     * If the environment variable BATTLECODE_GAME exists, clients should
     * set this field to its value.
     */
    gameID?: GameID;

    /**
     * The key to the gameID, if the gameID is key-protected.
     */
    key?: PlayerKey;
}

/**
 * Tells server to perform a list of action.
 */
export interface MakeTurn {
    command: "makeTurn";
    gameID: GameID;

    /**
     * Send the turn from the previous NextTurn received + 1.
     */
    turn: number;
    actions: Action[];
}

//////////////////////
// Server -> client //
//////////////////////

/**
 * Confirms that a player has been logged in.
 */
export interface LoginConfirm {
    command: "loginConfirm";
    gameID: GameID;
    teamID: TeamID;
}

/**
 * Notifies clients that a gameID has started.
 */
export interface GameStart {
    command: "start";
    gameID: GameID;
    initialState: GameState;
    teams: TeamData[];

    /**
     * Timeout for each turn.
     */
    timeoutMS: number;
}

/**
 * Notifies clients about the current turn / world state.
 */
export interface NextTurn {
    command: "nextTurn";
    gameID: GameID;
    turn: number;

    changed: EntityData[];
    dead: EntityID[];
    changedSectors: SectorData[];

    /**
     * The ID of the previous team that went.
     */
    lastTeamID: TeamID;
    successful: Action[];
    failed: Action[];
    reasons: string[];

    nextTeamID: TeamID;
    winnerID?: TeamID;
}

/**
 * Tells a team that it missed a turn.
 * Only sent to the team that missed; will be immediately followed by a NextTurn.
 */
export interface MissedTurn {
    command: "missedTurn";
    gameID: GameID;
    turn: number;
}

/**
 * Notifies client that its message was not understood.
 */
export interface ErrorCommand {
    command: "error";
    reason: string;
}

/**
 * Sends the entire state of the world, for testing purposes.
 */
export interface Keyframe {
    command: "keyframe";
    state: GameState;
    teams: TeamData[];
}

/**
 * Used by the viewer;
 */
export interface GameStatusUpdate {
    command: "gameStatusUpdate";
    gameID: GameID;
    // team names
    connected: string[];
    status: "lobby" | "running" | "finished" | "cancelled";
    map: string;
}

///////////////////
// Extra control //
///////////////////


// Note: client implementations in new languages need not implement the following commands.

/**
 * Create a gameID but do not participate.
 * At the end of the gameID, a GameData will be sent, containing all the information for the gameID.
 */
export interface CreateGame {
    command: "createGame";

    /**
     * The key to the server, if the server is password-protected.
     */
    serverKey?: string;

    /**
     * The teams that will be participating in the gameID.
     * Note: player .botName will override the .botName set here.
     * If keys are set, players will be assigned to teams based on keys.
     * Otherwise, players will be assigned in order of connection.
     */
    teams?: TeamData[];

    /**
     * The map the gameID will be played on.
     * Use a string to load a built-in map file with that name (case sensitive.)
     */
    map: MapFile | string;

    /**
     * States that the client is interested in receiving a replay when the game is complete.
     */
    sendReplay?: boolean;

    /**
     * Timeout for each turn, in milliseconds.
     * If you don't want a timeout, just set it really high.
     */
    timeoutMS?: number;
}

export interface CreateGameConfirm {
    command: "createGameConfirm";

    gameID: GameID;
}

export interface ListMapsRequest {
    command: "listMapsRequest";
}

export interface ListMapsResponse {
    command: "listMapsResponse";

    mapNames: string[];
    /**
     * Each entry in Maps is a MapFile encoded as JSON.
     * Woo, recursive encoding!
     */
    maps: string[];
}

export interface ListReplaysRequest {
    command: "listReplaysRequest";
}

export interface ListReplaysResponse {
    command: "listReplaysResponse";

    replayNames: string[];
}

export interface ReplayRequest {
    command: "replayRequest",
    name: string;
}

export interface ReplayResponse {
    command: "replayResponse",
    name: string;

    /**
     * A JSON-encoded MatchData.
     */
    match: string;
}

/**
 * Data sent to catch a client up to the current state of the gameID.
 * Optionally sent at the end of the gameID to whoever created it.
 */
export interface GameReplay {
    command: "gameReplay";

    // Note: this data is also included in the match data; these fields are provided
    // to make life simpler for the consumer
    id: GameID;

    winner?: TeamData;

    /**
     * The completed match.
     * A base64-encoded gzipped JSON-encoded MatchData.
     */
    matchData: string;
}

export interface PlayerConnected {
    command: "playerConnected",
    id: GameID;
    team: TeamID;
}

/**
 * Requests the server send all messages for all gameIDs.
 * Used by the viewer.
 */
export interface SpectateAll {
    command: "spectateAll";
}

/**
 * The data for a finished match.
 * Note that .bch17 files consist of a gzipped MatchData JSON object.
 */
export interface MatchData {
    /**
     * Required sanity check.
     */
    version: "battlecode 2017 hackathon match",

    gameID: GameID;
    initialState: GameState;
    winner?: TeamID;
    teams: TeamData[];
    turns: NextTurn[];
}

/**
 * Action types
 */
export type Action = MoveAction | PickupAction | ThrowAction | BuildAction | DisintegrateAction;

/**
 * Message types
 */
export type IncomingCommand = Login | MakeTurn | SpectateAll | CreateGame | ListMapsRequest |
    ListReplaysRequest | ReplayRequest;
export type OutgoingCommand = LoginConfirm | GameStart | NextTurn | MissedTurn | ErrorCommand | Keyframe |
    ListMapsResponse | ListReplaysResponse | GameReplay | ReplayResponse | PlayerConnected | CreateGameConfirm | GameStatusUpdate;

File: player-rust/src/schema.rs
pub type EntityID = u16;
pub type TeamID = u8;
pub type GameID = String;
pub type PlayerKey = String;

#[derive(PartialEq, Eq, Debug, Clone, Copy, Serialize, Deserialize)]
pub struct Location {
    pub x: i32,
    pub y: i32,
}

#[derive(PartialEq, Eq, Debug, Clone, Copy, Serialize, Deserialize)]
#[serde(rename_all="lowercase")]
pub enum EntityType {
    THROWER,
    STATUE,
    HEDGE
}

#[derive(PartialEq, Eq, Debug, Clone, Copy, Serialize, Deserialize)]
pub enum MapTile {
    #[serde(rename="G")]
    Grass,
    #[serde(rename="D")]
    Dirt
}

#[derive(PartialEq, Eq, Debug, Clone, Copy, Serialize, Deserialize)]
#[serde(rename_all="camelCase")]
pub struct EntityData {
    pub id: EntityID,
    // note: this field is renamed because "type" is a keyword in rust
    #[serde(rename="type")]
    pub entity_type: EntityType,
    pub location: Location,
    #[serde(rename="teamID")]
    pub team_id: TeamID,
    pub hp: u8,
    pub cooldown_end: Option<u32>,
    pub held_by: Option<EntityID>,
    pub holding: Option<EntityID>,
    pub holding_end: Option<u32>,
}

#[derive(PartialEq, Eq, Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all="camelCase")]
pub struct SectorData {
    top_left: Location,
    #[serde(rename="controllingTeamID")]
    controlling_team_id: TeamID
}

#[derive(PartialEq, Eq, Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all="camelCase")]
pub struct GameState {
    pub version: String,
    pub height: u16,
    pub width: u16,
    /// Indexed as [y][x]
    pub tiles: Vec<Vec<MapTile>>,
    pub sector_size: u16,
    pub entities: Vec<EntityData>,
    pub sectors: Vec<SectorData>
}

#[derive(PartialEq, Eq, Debug, Clone, Serialize, Deserialize)]
pub struct TeamData {
    pub id: TeamID,
    pub name: String,
    pub key: Option<PlayerKey>
}

#[derive(PartialEq, Eq, Debug, Serialize, Deserialize)]
#[serde(tag = "command")]
#[serde(rename_all = "camelCase")]
pub enum CommandToServer {
    Login {
        name: String,
    },
    MakeTurn {
        turn: u32,
        actions: Vec<Action>,
    }
}

#[derive(PartialEq, Eq, Debug, Serialize, Deserialize)]
#[serde(tag = "command")]
#[serde(rename_all = "camelCase")]
pub enum CommandToClient {
    LoginConfirm {
        #[serde(rename="teamID")]
        game_id: GameID,

        name: String,
        #[serde(rename="teamID")]
        team_id: TeamID,
    },
    Start {
        #[serde(rename="teamID")]
        game_id: GameID,

        map: Map,
        teams: Vec<TeamData>,
    },
    #[serde(rename_all="camelCase")]
    NextTurn {
        #[serde(rename="teamID")]
        game_id: GameID,
        turn: u32,

        changed: Vec<EntityData>,
        dead: Vec<EntityID>,
        changed_sectors: Vec<SectorData>,

        #[serde(rename="lastTeamID")]
        last_team_id: TeamID,
        successful: Option<Vec<Action>>,
        failed: Option<Vec<Action>>,
        reasons: Option<Vec<String>>,

        #[serde(rename="nextTeamID")]
        next_team_id: TeamID,
        #[serde(rename="winnerID")]
        winner_id: Option<TeamID>,
    },
    Error {
        reason: String
    },
    Keyframe {
        state: GameState,
        teams: Vec<TeamData>
    }
}

#[derive(PartialEq, Eq, Debug, Serialize, Deserialize)]
#[serde(tag = "action")]
#[serde(rename_all = "camelCase")]
pub enum Action {
    Move {
        id: EntityID,
        dx: i8,
        dy: i8
    },
    Pickup {
        id: EntityID,
        pickupid: EntityID,
    },
    Throw {
        id: EntityID,
        dx: i8,
        dy: i8
    },
    Build {
        id: EntityID,
        dx: i8,
        dy: i8
    },
    Disintegrate {
        id: EntityID,
    }
}

#[cfg(test)]
mod tests {
    use serde_json;
    use super::*;

    #[test]
    fn test_deserialization() {
        let login: CommandToServer = serde_json::from_str(r#"
        {
            "command": "login",
            "name": "test"
        }
        "#).unwrap();

        assert_eq!(login, CommandToServer::Login { name: "test".into() });

        let make_turn: CommandToServer = serde_json::from_str(r#"
        {
            "command": "makeTurn",
            "turn": 0,
            "actions": [
                {
                    "action": "move",
                    "id": 57,
                    "loc": {"x": 0, "y": 20}
                },
                {
                    "action": "pickup",
                    "id": 57,
                    "pickupid": 78
                },
                {
                    "action": "throw",
                    "id": 57,
                    "loc": {"x": 0, "y": 20}
                },
                {
                    "action": "build",
                    "id": 57,
                    "loc": {"x": 0, "y": 20}
                },
                {
                    "action": "disintegrate",
                    "id": 57
                }
            ]
        }
        "#).unwrap();

        assert_eq!(make_turn, CommandToServer::MakeTurn {
            turn: 0,
            actions: vec![
                Action::Move {
                    id: 57,
                    loc: Location {x: 0, y: 20}
                },
                Action::Pickup {
                    id: 57,
                    pickupid: 78
                },
                Action::Throw {
                    id: 57,
                    loc: Location {x: 0, y: 20}
                },
                Action::Build {
                    id: 57,
                    loc: Location {x: 0, y: 20}
                },
                Action::Disintegrate {
                    id: 57
                }
            ]
        });

        let login_confirm: CommandToClient = serde_json::from_str(r#"
            {
                "command": "loginConfirm",
                "name": "test",
                "id": 0
            }
        "#).unwrap();
        assert_eq!(login_confirm, CommandToClient::LoginConfirm {
            name: "test".into(),
            id: 0
        });

        let start: CommandToClient = serde_json::from_str(r#"
        {
            "command": "start",
            "map": {
                "width": 2,
                "height": 2,
                "tiles": [["G","D"],["D","G"]]
            },
            "teams": [
                {
                    "id": 0,
                    "name": "test"
                }
            ]
        }
        "#).unwrap();
        assert_eq!(start, CommandToClient::Start {
            map: Map {
                width: 2,
                height: 2,
                tiles: vec![vec![MapTile::Grass, MapTile::Dirt],
                            vec![MapTile::Dirt, MapTile::Grass]]
            },
            teams: vec![
                Team {
                    id: 0,
                    name: "test".into()
                }
            ]
        });
        
        let next_turn: CommandToClient = serde_json::from_str(r#"
        {
            "command": "nextTurn",
            "turn": 0,
            "changed": [{
                 "id": 0,
                 "type": "thrower",
                 "location": {"x": 0, "y": 0},
                 "team": 0,
                 "hp": 255,
                 "cooldownEnd": 822,
                 "heldBy": 3,
                 "holding": 5,
                 "holdingEnd": 993
            }],
            "dead": [1, 2],
            "successful": [{"action": "disintegrate", "id": 75}],
            "failed": [{"action":"disintegrate", "id": 2388}],
            "reasons": ["bot does not exist: 2388"],

            "nextTeam": 3,
            "winner": 7
        }
        "#).unwrap();

        assert_eq!(next_turn, CommandToClient::NextTurn {
            turn: 0,
            changed: vec![Entity {
                 id: 0,
                 entity_type: EntityType::Thrower,
                 location: Location {x: 0, y: 0},
                 team: 0,
                 hp: 255,
                 cooldown_end: Some(822),
                 held_by: Some(3),
                 holding: Some(5),
                 holding_end: Some(993)
            }],
            dead: vec![1, 2],
            successful: Some(vec![Action::Disintegrate {id: 75}]),
            failed: Some(vec![Action::Disintegrate {id: 2388}]),
            reasons: Some(vec!["bot does not exist: 2388".into()]),

            next_team: 3,
            winner: Some(7)
        });
    }
}